Add "ViewLayout" prop to liquid / razor page - Add "layout" liquid tag.

jtkech opened the pull request:
Fixes #1425 

Then, i could define a custom layout just by doing this in a liquid view.

    {% no_render ThemeLayout | add_alternates: "Layout2" %}

---

@sebastienros just saw your comment on the related issue, so i will create a dedicated tag.

> We need a tag for liquid. Like the standard layout tag in liquid.

Okay

> And a property in the base class that when assigned will insert an alternate at the top

Why? Through a tag we can already easily add an alternate to the `ThemeLayout` ambient value. And maybe in the future we would want to `clear_alternates` and `add_classes` ... So, maybe also good to still add `ThemeLayout` in the context local scope, and have a `no_render` tag (or another name).

> We also need to ensure that this alternate is not actually preserved across requests, I am not really sure of it.

As i have tried, this is not the case.

---

**Update**

- I also added shape `Wrappers` filters.

- I added a `ViewLayout` property to our `RazorPage` class. Anyway, in razor views it will be easier to just do `ViewLayout = "SomeLayout"`. Didn't test it in a razor view, i will check it tomorrow.

- Then, i wanted to use this property in the liquid tag, but currently our liquid page inherit from the mvc razor page, not from our custom one. Before it was the case but i changed it when we harmonized `Contextualize()` called from different contexts and began to pass the `DisplayContext` in place of the `LiquidPage`, and then resolve services ourself.

- So here, the `LiquidPage` now inherits from our custom `RazorPage`. Then, because i can retrieve it from the `DisplayContext`, i set it in the ambient values and then set the new `ViewLayout` prop in the tag.

- Note: See in the code, i kept both versions of the `layout` tag, one which use the new `ViewLayout` property and one which update directly `ThemeLayout.Metadata.Alternates`.

sebastienros commented on src/OrchardCore/OrchardCore.DisplayManagement.Liquid/Tags/LayoutTag.cs:
```diff
@@ -0,0 +1,35 @@
+            if (page != null && !string.IsNullOrWhiteSpace(name))
+            {
+                // So we only check if the page has a 'ViewLayout' property.
+                if (((object)page).GetType().GetProperty("ViewLayout") != null)
```
It's defined in RazorPage, it should be a a base class then, no generics (if this is the issue) and you can create a Dynamic Delegate that is cached and sets the property on this page instance instead of invoking reflection all the time.

jtkech replied:
@sebastienros 

Here it could be a `LiquidPage : Razor.RazorPage<dynamic>` , so okay., Then when coming e.g from a db template, we don't use a custom file provider => razor engine => bridge to liquid engine. So, in this case it could be our `Razor.RazorPage<dynamic>` or a `Razor.RazorPage<SomeOtherViewModelProxy>` so i was not able to cast to the right type without using reflexion.

Hmm, maybe just because we define in `RazorPage<TModel>` all properties (which don't use `<TModel>`), in place of defining these properties in a non generic base class. I will check this.

sebastienros commented on src/OrchardCore/OrchardCore.DisplayManagement/Razor/RazorPage.cs:
```diff
@@ -119,6 +120,36 @@ public dynamic ThemeLayout
+            {
+                if (ThemeLayout is IShape layout)
+                {
+                    var alternates = layout.Metadata.Alternates.ToList();
```
Too many new collection to insert an element. If the `AlternatesColletion` doesn't have a way to insert an item, let's add it.

jtkech replied:
Okay, i will do.

jtkech replied:
Done

jtkech commented:
@sebastienros 

So, actually we have the following

    public class LiquidPage : Razor.RazorPage<dynamic>
    public abstract class RazorPage : RazorPage<dynamic>
    public abstract class RazorPage<TModel> : Microsoft.AspNetCore.Mvc.Razor.RazorPage<TModel>

I tried, without reflexion, in the context of an editor shape **with a custom model, here the page doesn't cast to `RazorPage` neither `RazorPage<dynamic>`**. So i tried different definitions, e.g.

    public abstract class RazorPage : AspNetCore.Mvc.Razor.RazorPage
    { here our custom implementation }

    // But here we can't inherit of 2 classes with the same base class 'RazorPageBase'.
    public abstract class RazorPage<TModel> : AspNetCore.Mvc.Razor.RazorPage<TModel>, RazorPage

**But no problem** in all cases the page is a `RazorPageBase`, so, because it is intended to be used in a `LiquidPage` or our `Razor.RazorPage<SomeModelType>`, now i just retrieve the page in a dynamic var, just check if it is a `RazorPageBase` and then set the `ViewLayout` property.

sebastienros commented:
Why do we need a `no_render` now and not before? `add_alternates` was already available.

jtkech commented:
it allows to alter a shape without rendering it as i was first using to not have to create the `layout` tag.

    {% no_render ThemeLayout | add_alternates: "Layout2" %}

This to not have to use e.g the `assign` tag.

    {% assign test = ThemeLayout | add_alternates: "Layout2" %}

Anyway, we had the `layout` tag to insert the alternate at the first position. Humm, now maybe good to also create an `insert_alternate` filter.

**But i can remove the `no_render` tag if you want** or maybe a more general one just to allow filters to execute without a specific tag. Hmm, maybe better to add this to the syntax so that it is done automatically when no tag is specified, a kind of nop tag but which let filters evaluate.

sebastienros commented:
Not sure the grammar could detect there is no tag, and it's going away from liquid's syntax.
So it's a custom tag to prevent using another one, `assign`. Then it's just that `add_alternate` should be a tag and not a filter.

So my recommendation would be to change any filter whose result can't be reused into a tag. no need for insert alternate. I'd have to check but I think that most precise alternates are supposed to be at the end of the list, and the shape display logic starts from the end. So Add is supposed to be sufficient. If this is the case then the change to AlternatesCollection is also wrong.

jtkech commented:
Okay.

So, i will remove `no_render` and in the setter of the `ViewLayout` property i will replace "Insert" by "Add". I think you're right because of this.

            foreach (var shapeAlternate in shapeAlternates.Reverse())

I knew this, maybe because i totally trusted you when you asked me to do it ;)

Then, do i need to remove the new `Insert()` method in the `AlternatesCollection`?

Then i suggest to let the `add_alternates` as a filter because it's more useful to be able to do.

    {% display Model | add_alternates "***" | shape_type: "***" %}

sebastienros commented:
Remove the inserts if they are not needed.

About filter, not sure. It's easy for your functional oriented mind. Might be easier to just use tags for a standard user:

```
{% shape_alternate Model "alternate" %}
{% shape_type Model "alternate" %}
{% display Model %}
```

jtkech commented:
Okay, no problem i will make them tags.

Hmm, it's a little breaking change for existing sites which use the dev branch.

jtkech commented:
@sebastienros i removed the alternates insert and the no_render tag.

Then, about moving filters to tags, i realized that i would need to move about 15 filters. I have no  closed opinion but **I need your confirmation**.

So,

    {% display Model | clear_alternates | shape_type: "MenuItemLink" %}

Would become

    {% clear_alternates Model.Content %}
    {% shape_type Model.Content, "MenuItemLink" %}
    {% display Model.Content" %}

And the same for most of filters defined in `.DisplayManagement.Liquid`, e.g.

    {% display Model.Content | remove_item: "BodyPart" %}

Would become

    {% remove_item Model.Content, "BodyPart" %}
    {% display Model.Content" %}

sebastienros commented:
@paynecrl97 we updated most shape filters to become tags, and also changed their names such they are prefixed by `shape`. 

This means `{% display Model.Content %}` for instance now becomes `{% shape_display Model.Content %}`. We think it's clearer what the tags do, some users being confused and using content items after `display`.

We want to get your opinion before merging it. Look at the changes in the README to see what is impacted.

paynecrl97 commented:
I'll pull down at some point in the next couple of days and give it a go

jtkech commented:
@sebastienros 

- I added here the `shape_pager` tag and removed the `set_properties` tag. I did it here because i needed the new `ExpressionArgumentsTag` base class.

- I saw another issue on the pager usage in the list part driver. The main problem is that when we use in a query an `OrderBy[Descending]()` then a `Take()`, the ordering is lost. Maybe because of this in yessql.


            IQuery<T> IQuery<T>.Take(int count)
            {
                _query._sqlBuilder.ClearOrder(); <==
                _query._sqlBuilder.OrderBy(_query._dialect.QuoteForColumnName("Id"));
                _query._sqlBuilder.Take(count.ToString());
                return this;
            }

    So, in the first rendering we can't show correctly the n last created posts (tested with `TheBlogTheme`), assuming (and it seems to be) that blog posts in the db keep ordered by created date (ascending order).

    **So here, just to show, can be easily reverted**, we first do a `CountAsync()` so that in a 2nd query we can skip `count - page size` to grab the n last (= page size) blog posts.

    Then, i removed all `OrderBy[Descending]()` in each query which uses `Take()` and rather did the ordering on the returned collection.

sebastienros commented:
Please revert the changes for the order by, you are actually breaking the query altogether as the Take() is not ordered, and it won't work when there are more items than the page size. It's a bug is YesSql that I will fix.

jtkech commented:
Okay, no problem it was just a temporary workaround to show the issue, i will revert it.

To make it working, the first time (without after / before), i was doing a `CountAsync()` and then a `Skip(count - (page size + 1))` to grab the n last items, so it was working even there are more items than the page size. And in all cases, i was ordering things after the query on the returned collection.

jtkech commented:
@sebastienros 

So, here what i just did but i can't commit it because of a network issue with github.

- Revert the list part workaround.

- Rename "build_display" to "shape_build_display" and updated the doc.

- In the doc, concatenate 2 sections of "Shape Tags" into only one.

- Maybe i will have a little conflict because i updated a file which was moved (to put 2 liquid filters in the same folder) in another merged PR, but not a problem at all.

sebastienros commented:
Update this https://github.com/Microsoft/Git-Credential-Manager-for-Windows/releases/tag/v1.14.0

jtkech commented:
@sebastienros thanks, just done and i also needed to update git, and my git extensions settings.

sebastienros commented:
Dev branch has the fix for the paging and order

jtkech commented:
Okay cool

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Liquid/README.md:
````diff
@@ -250,172 +250,213 @@ Output
 ```
 
-## Layout Tags
+Whenever the type is changed, it is recommended to clear the shape alternates before using the `shape_clear_alternates` tag.
````
Right, there is no correct order. We could also do it automatically when this tag is called.

sebastienros commented:
@paynecrl97 I looked at all the usages of liquid filters, and__ none are mutating the state on the input__, this is counter intuitive in liquid. In some cases with shapes I understand it's easier to just chain "actions" but it's easier for you because you are not an average developer. So I assume the designer of liquid made it intentional. This flaw in our filters showed up when we had to create a custom `{% no_render %}` that would just do nothing with a shape, as there was no tag to alter a state without rendering anything.

Also, it's not very common to have such constructs. We already replaced the usage in the blog theme, and in the end I found only two places where we changed a single line into multiple ones:

```
{% display Model.Content | remove_item: "BodyPart" %}
```
is replaced by 
```
{% shape_remove_item Model.Content "BodyPart" %}
{% shape_display Model.Content %} 
```

I believe it's easier to explain the second construct to an editor than the first one.

Do these arguments make you see things differently? I don't want to force any decision.

paynecrl97 commented:
I was expecting the chained construct to be used much more frequently than once (I was thinking along the lines of menu shapes- it's quite common to alter menu shapes e.g. one menu may be horizontal, one may be stacked etc...).

As for liquid not being designed like this, the liquid docs state that `Filters change the output of a Liquid object` (https://shopify.github.io/liquid/basics/introduction/#filters). This to me reads that filters are responsible for altering objects and data (eg `upcase`, `round`, reverse` etc..). Tags `create the logic and control flow`.

I accept that a `no_render` tag (as currently would be required) is not ideal. Maybe there's another approach/ compromise here?

paynecrl97 commented:
Ah OK, I see your point.

But also there are no Liquid tags that edit the state of an object. The closest is `assign`, which can be used in tandem with filters to modify an object and save it for later use if required. eg: `{% assign x = x | upcase %} {{ x }} {{ x }}` would produce `FOO FOO`.

If filters are concerned with output, then `shape_display` shouldn't be a tag. The example you gave above (`{% display Model.Content | remove_item: "BodyPart" %}`), would become `{{ Model.Content | shape_remove_item: "BodyPart" | shape_build_display }}`. This (to me) is more natural as we're starting with an object, massaging it into a state we require, then writing it to output.

paynecrl97 commented:
So I just had a look at our usages of tags, and it appears that most (if not all) write to the output. I can't find any other implementations/extensions of liquid that do this.

I'm also not convinced that we should be using filters to retrieve data (`query` and `content` filters). Shopify and Jekyll both do this with scoped values. For example, `{% assign posts = "LatestBlogPosts" | query %}` could become `{% assign posts = Queries.LatestBlogPosts %}`, and to retrieve content could be something like `{{ Content.MyContentAlias | display_text }}`.

sebastienros commented:
Retrieving data using filters, yes I agree, and I did it at first.  For content by alias yes, that's better. The issue I face is passing arguments. You can't pass arguments to `Query.MyQuery`, or we need to construct a parameter object as the input for this query using another variable first.
I will look for examples.

For chaining, I assume the example you wanted to provide is `{{ Model.Content | shape_remove_item: "BodyPart" | shape_render }}` instead of `shape_build_display` which would take a content item and create a shape. Not that `shape_render` is current called `shape_stringify` in this PR.
Maybe a solution would be to have a term be both a filter and a tag, like `raw` is, for some statements. In the case of `shape_display` as a tag, I would agree to remove it if we have a filter, and I also agree that using a filter to output some text is better. So my example would then be

```
{% shape_remove_item Model.Content "BodyPart" %}
{% shape_remove_item Model.Content "TitlePart" %}
{{ Model.Content | shape_render }} 
```

Or with your vision, I see two alternatives

```
{{ Model.Content | shape_remove_item: "BodyPart" | shape_remove_item: "TitlePart" | shape_render }}
```

or 

```
{{ Model.Content | shape_remove_item: "BodyPart", "TitlePart" | shape_build_display }}
```

I would be tempted to suggest that these filters actually return a new shape, so they are not altering their states. That would sadden me as it would make more work on the server, but not much as it's a rare construct.

paynecrl97 commented:
Yes, passing arguments into a query would need to be solved. As would retrieving content by ID as you can currently do with the `content` filter (I'm guessing that `Content.<Some ID>` would not be valid as the ID would be numeric). Although we could consider only allowing content with aliases to be retrieved in this way.

I've read those examples above over and over and I can't decide which would be better. I think the filters read more naturally and are also closer to expected Liquid behaviour **but only if the filters return new shapes so that they don't alter the original object's state** (and I know this would sadden you).

If we decide that we can't take that perf hit then I think we should stick to tags.

We've kind of strayed away from the contents of this PR a bit, is it worthwhile me creating a new project for this with tickets for each potential change? That way we can discuss and track them all independently of each other.

sebastienros commented:
We can use another PR for Query and Content. I want to solve the filter/tag issues here as I don't want to break it multiple times.

Let's agree on immutable filters that would not change the shape but return a new one. I will go over it again with @jtkech.

paynecrl97 commented:
Sounds good to me

Sent from my iPhone

> On 26 Feb 2018, at 21:58, Sébastien Ros <notifications@github.com> wrote:
> 
> We can use another PR for Query and Content. I want to solve the filter/tag issues here as I don't want to break it multiple times.
> 
> Let's agree on immutable filters that would not change the shape but return a new one. I will go over it again with @jtkech.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

sebastienros commented:
After looking into with @jtkech it appears we can't clone the shapes. They are from poco classes, not just the `Shape` class.

> but only if the filters return new shapes so that they don't alter the original object's state

So right now there are some things that can go back to filters, like `shape_render` and some others that can't.

In the end we can still do the filter syntax for the most common cases which would be like

- `{{ Model.ContentItem | shape_build_display: "Summary" | shape_render }}`
- `{{ Model.Content | shape_render }}`
- `{{ Model.Content.BodyPart | shape_render }}`

And only in the case of having to alter a shape use the specific tags, that won't render anything and only alter the state.

paynecrl97 commented:
Added #1532 and #1533 to track content and queries changes
